symengine/functions.cpp

```diff
@@ -329,6 +329,118 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
+            return false;
+        }
+    }
+    if (is_a<Pow>(*arg)) {
+        if (is_a_Complex(*down_cast<const Pow &>(*arg).get_base())
+            and down_cast<const ComplexBase &>(
+                    *down_cast<const Pow &>(*arg).get_base())
+                    .is_re_zero()) {
+            if (is_a<Integer>(*down_cast<const Pow &>(*arg).get_exp())) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+RCP<const Basic> Sign::create(const RCP<const Basic> &arg) const
+{
+    return sign(arg);
+}
+
+RCP<const Basic> sign(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (is_a<NaN>(*arg)) {
+            return Nan;
+        }
+        if (down_cast<const Number &>(*arg).is_zero()) {
+            return zero;
+        }
+        if (down_cast<const Number &>(*arg).is_positive()) {
+            return one;
+        }
+        if (down_cast<const Number &>(*arg).is_negative()) {
+            return minus_one;
+        }
+        if (is_a_Complex(*arg)
+            and down_cast<const ComplexBase &>(*arg).is_re_zero()) {
+            RCP<const Basic> r
+                = down_cast<const ComplexBase &>(*arg).imaginary_part();
+            if (is_a_Number(*r)
+                and down_cast<const Number &>(*r).is_positive()) {
+                return I;
+            }
+            if (is_a_Number(*r)
+                and down_cast<const Number &>(*r).is_negative()) {
+                return mul(minus_one, I);
+            }
+        }
+    }
+    if (is_a<Constant>(*arg)) {
+        if (eq(*arg, *pi) or eq(*arg, *E) or eq(*arg, *EulerGamma)
+            or eq(*arg, *Catalan) or eq(*arg, *GoldenRatio))
+            return one;
+    }
+    if (is_a<Sign>(*arg)) {
+        return arg;
+    }
+    if (is_a<Mul>(*arg)) {
+        RCP<const Basic> s = sign(down_cast<const Mul &>(*arg).get_coef());
+        map_basic_basic dict = down_cast<const Mul &>(*arg).get_dict();
+        RCP<const Number> coef = one;
+        map_basic_basic new_dict;
+        for (const auto &p : dict) {
+            if (is_a_Complex(*p.first)
+                and down_cast<const ComplexBase &>(*p.first).is_re_zero()) {
+                if (is_a<Integer>(*p.second)) {
+                    s = mul(s, pow(sign(p.first), p.second));
+                }
+            } else {
+                Mul::dict_add_term_new(outArg(coef), new_dict, p.second,
+                                       p.first);
+            }
+        }
+        return mul(
+            s, make_rcp<const Sign>(Mul::from_dict(coef, std::move(new_dict))));
+    }
+    if (is_a<Pow>(*arg)) {
+        RCP<const Basic> base = down_cast<const Pow &>(*arg).get_base();
+        RCP<const Basic> exp = down_cast<const Pow &>(*arg).get_exp();
+        if ((is_a_Complex(*base)
+             and down_cast<const ComplexBase &>(*base).is_re_zero())
+            and is_a<Integer>(*exp)) {
+            return pow(sign(base), exp);
```

Comment:
Are you sure this is true? You have to be extra careful with complex numbers. Since SymPy doesn't do it, I wouldn't do this either.